src/mqtt_client.rs

```diff
@@ -536,8 +571,47 @@ impl<
                         self.network.send_packet(&puback)?;
                     }
 
-                    // TODO: Add support.
-                    QoS::ExactlyOnce => unimplemented!(),
+                    QoS::ExactlyOnce => {
+                        // Note(uwnrap): There should always be a packet ID for QoS >
+                        // AtMostOnce.
+                        let packet_id = info.packet_id.unwrap();
```

Comment:
Doesn't this come from the broker?
We should never panic on erroneous broker data. Either handle, ignore, or reset.

Reply:
This doesn't come from the broker, it comes from our deserializer. The QoS bits of the control byte tell us whether or not the ID is present in the message. So if QoS > AtMostOnce, there's always a packet_id in the option.

I agree we should never panic based on external input